src/nfa/operations.cc

```diff
@@ -114,6 +114,364 @@ namespace {
+                            result.delta.add(Sid, currentSymbol, switch_target);
+                    }
+                }
+            }
+        }
+
+        delete subset_map;
+        return result;
+    }
+
+    void residual_recurse_coverable(const std::vector <StateSet>& macrostate_vec,   // vector of nfa macrostates
+                                    const std::vector <State>& covering_indexes,    // sub-vector of macrostates indexes
+                                    std::vector <bool>& covered,                    // flags of covered states
+                                    std::vector <bool>& visited,                    // flags fo visited states
+                                    unsigned long start_index,                      // starting index for covering_indexes vec
+                                    std::unordered_map<StateSet, State> *subset_map,    // mapping of indexes to macrostates
+                                    Nfa& nfa) {
+
+        StateSet check_state = macrostate_vec[covering_indexes[start_index]];
+        StateSet covering_set;                      // doesn't contain duplicates
+        std::vector<State> sub_covering_indexes;    // // indexes of covering states
+
+        for (auto i = covering_indexes.begin() + static_cast<long int>(start_index+1), e = covering_indexes.end(); i != e; i++) {
+            if (covered[*i])           // was aready processed
+                continue;
+
+            if (macrostate_vec[*i].IsSubsetOf(check_state)) {
+                covering_set.insert(macrostate_vec[*i]);                // is never covered
+                sub_covering_indexes.push_back(*i);
+            }
+        }
+
+        if (covering_set == check_state) {       // can recurse even without covered :thinking:
+
+            for (unsigned long k = 0; k < sub_covering_indexes.size()-1; k++) {
+                if (macrostate_vec[sub_covering_indexes[k]].size() == 1)            // end on single-sized states
+                    break;
+
+                if (visited[sub_covering_indexes[k]])                               // already processed
+                    continue;
+
+                visited[sub_covering_indexes[k]] = true;
+
+                residual_recurse_coverable(macrostate_vec, sub_covering_indexes, covered, visited, k, subset_map, nfa);
+            }
+
+            covering_set.clear();                 // clear variable to store only needed macrostates
+            for (auto index : sub_covering_indexes) {
+                if (covered[index] == 0) {
+                    auto macrostate_ptr = subset_map->find(macrostate_vec[index]);
+                        if (macrostate_ptr == subset_map->end())        // should never happen
+                             throw std::runtime_error(std::to_string(__func__) + " couldn't find expected element in a map.");
+
+                    covering_set.insert(macrostate_ptr->second);        //todo check for null but should be impossible
```

Comment:
Could we try to resolve this TODO before merging? If not, at least rename the todo to `// TODO: <message>` so that it is easily searchable and recognized by IDEs and such.

Reply:
Just forgot to remove the comment, the todo was already resoled as it was to check whether the `macrostate_ptr` is a valid pointer to `subset_map` , which should always find the element in the map.

Reply:
Great. Feel free to close this conversation as resolved.